### PR TITLE
Parallelization of test cases

### PR DESCRIPTION
**Pull request by TheAnkurGoswami**
## What does this PR do?

Parallelize tests on github actions to utilize all available nodes.

## Before reviewing

- [ ] Did you test this PR?
- [ ] Is there any migration or seeding necessary?
- [ ] If a new field is added to signals in the api, did you update the sample test files?

### Links:

- Jira: https://myaccern.atlassian.net/browse/RSL-523

### CODE REVIEW

File: .github/workflows/python-app.yml [hunk 1]
```diff
@@ -105,15 +115,87 @@ jobs:
       run: |
         make lint-type-check
 
-    - name: Run Tests Coverage
+    - name: Run test on Node ${{ matrix.ci_node_index }}
+      run: |
+        set -e
+        mkdir -p test-results
+        rm -r test-results/parts || echo "no files to delete"
+        OUT=$(python -c "from tests.utils import split_tests; split_tests('test-results/results.xml', ${{ matrix.ci_node_total }}, ${{ matrix.ci_node_index }})")
```
Comment (JosuaKrause) [inline .github/workflows/python-app.yml:123]: add a `__main__` and actual command line arguments

File: tests/utils.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                test_time_map[f"{classname}{PY_EXT}"] += float(
+                    testcases.attrib["time"])
+
+            for file in test_files:
+                if file not in test_time_map.keys():
+                    test_time_map[file] = DEFAULT_TEST_DURATION
+
+            time_keys: List[Tuple[str, float]] = sorted(
+                test_time_map.items(), key=lambda el: el[1], reverse=True)
+        except FileNotFoundError:
```
Comment (JosuaKrause) [inline tests/utils.py:80]: do that in the loop -- other files might still be there?

File: tests/utils.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            time_keys = [(file, DEFAULT_TEST_DURATION) for file in test_files]
+
+        def find_lowest_total_time(
+                    test_sets: List[Tuple[List[str], float]]) -> int:
```
Comment (JosuaKrause) [inline tests/utils.py:84]: indent

File: tests/utils.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            for ix, val in enumerate(test_sets):
+                if minimum is None or val[1] < minimum:
+                    minimum = val[1]
+                    idx = ix
+            return idx
+
+        test_sets: List[Tuple[List[str], float]] = [
+            ([], 0.0) for _ in range(total_nodes)]
+        for key, timing in time_keys:
+            ix = find_lowest_total_time(test_sets)
```
Comment (JosuaKrause) [inline tests/utils.py:96]: why are you finding the lowest time in test_sets? should be time_keys

File: .github/workflows/python-app.yml [hunk 5]
```diff
@@ -87,7 +89,15 @@ jobs:
       with:
         python-version: ${{ matrix.python }}
 
-    - name: Cache python modules ${{ matrix.python }}
+    - name: Restore test-result Cache
+      uses: actions/cache@v3
+      with:
+        path: test-results/*
+        key:  v1-test-results-${{ matrix.python }}  # meant to be wrong so that restore-keys can find it
```
Comment (JosuaKrause) [inline .github/workflows/python-app.yml:96]: why does the key need to be wrong?

File: .github/workflows/python-app.yml [hunk 6]
```diff
@@ -105,15 +115,87 @@ jobs:
...
+        IFS=',' read -a FILE_INFO <<< "$OUT"
+        FILES=("${FILE_INFO[@]:1}")
+        echo "Selected Files are: " ${FILES[@]}
+        echo "Previous timings are: ${FILE_INFO[0]}"
+        ./coverage/coverage.sh results-${{ matrix.python }}-${{ matrix.ci_node_index }}.xml ${FILES[@]}
+
+    - name: Upload Artifact(results)
+      uses: actions/upload-artifact@v3
+      with:
+        name: results-${{ matrix.python }}-${{ matrix.ci_node_index }}
```
Comment (JosuaKrause) [inline .github/workflows/python-app.yml:133]: this one is even more different than above

File: .github/workflows/python-app.yml [hunk 7]
```diff
@@ -105,15 +115,87 @@ jobs:
...
+    - name: Set up Python ${{ matrix.python }}
+      uses: actions/setup-python@v1
+      with:
+        python-version: ${{ matrix.python }}
+
+    - name: Cache python modules
+      uses: actions/cache@v2
+      with:
+        path: ${{ env.pythonLocation }}
+        key: ${{ env.pythonLocation }}-${{ hashFiles('**/requirements.txt') }}-${{ hashFiles('**/requirements.lint.txt') }}
```
Comment (JosuaKrause) [inline .github/workflows/python-app.yml:169]: why **?

File: .github/workflows/python-app.yml (hunk 5 above)
Comment (JosuaKrause) [reply .github/workflows/python-app.yml:96]: also use hashFiles on tests

File: tests/utils.py (hunk 4 above)
Comment (TheAnkurGoswami) [reply tests/utils.py:96]: the bucket which has less sum/total gets next addon.

File: .github/workflows/python-app.yml (hunk 6 above)
Comment (TheAnkurGoswami) [reply .github/workflows/python-app.yml:133]: This is just for artifact. Once workflow completes, we cannot access artifact using keys. We can use artifacts inside the workflow only.

The key above is just to pass information from `tests` to `coverage`.

File: .github/workflows/python-app.yml (hunk 7 above)
Comment (TheAnkurGoswami) [reply .github/workflows/python-app.yml:169]: got that from xyme-backend. will change.

File: tests/utils.py (hunk 2 above)
Comment (TheAnkurGoswami) [reply tests/utils.py:80]: No, that is for the case when we don;t have timings xml file. We want default time for each test case.

File: .github/workflows/python-app.yml [hunk 8]
```diff
@@ -105,15 +115,87 @@ jobs:
...
         make lint-type-check
 
-    - name: Run Tests Coverage
+    - name: Run test on Node ${{ matrix.ci_node_index }}
       run: |
-        make coverage-report
+        set -e
+        mkdir -p test-results
+        rm -r test-results/parts || echo "no files to delete"
+        OUT=$(python -m tests --filepath test-results/results.xml --total-nodes ${{ matrix.ci_node_total }} --node-id ${{ matrix.ci_node_index }})
```
Comment (JosuaKrause) [inline .github/workflows/python-app.yml:123]: where are you combining the result files?

File: .github/workflows/python-app.yml [hunk 9]
```diff
@@ -87,11 +89,19 @@ jobs:
       with:
         python-version: ${{ matrix.python }}
 
-    - name: Cache python modules ${{ matrix.python }}
-      uses: actions/cache@v2
+    - name: Restore test-result Cache
+      uses: actions/cache@v3
+      with:
+        path: test-results/*
+        key:  v1-test-results-${{ matrix.python }}-${{ hashFiles('tests/test*.py')}}-${{ hashFiles('tests/data*') }}
```
Comment (JosuaKrause) [inline .github/workflows/python-app.yml:96]: `tests/data/**`

File: .github/workflows/python-app.yml [hunk 10]
```diff
@@ -20,10 +20,10 @@ jobs:
         python-version: ${{ matrix.python }}
 
     - name: Cache python modules
-      uses: actions/cache@v2
+      uses: actions/cache@v3
       with:
         path: ${{ env.pythonLocation }}
-        key: ${{ env.pythonLocation }}-${{ hashFiles('**/requirements.txt') }}-${{ hashFiles('**/requirements.lint.txt') }}
+        key: ${{ env.pythonLocation }}-${{ hashFiles('requirements.txt') }}-${{ hashFiles('requirements.lint.txt') }}
```
Comment (JosuaKrause) [inline .github/workflows/python-app.yml:26]: make sure the python version is accounted for

File: .github/workflows/python-app.yml [hunk 11]
```diff
@@ -105,15 +115,87 @@ jobs:
...
+        make install
+        mkdir -p test-results/parts
+
+    - name: Download all results
+      # Downloads results-*.xml,
+      uses: actions/download-artifact@v2
+    - name: Merge results
+      run: |
+        mv results-${{ matrix.python }}-*/results-${{ matrix.python }}-* test-results/parts/
+        python3 -c "from tests.utils import merge_results; merge_results('./test-results', 'results.xml')"
```
Comment (JosuaKrause) [inline .github/workflows/python-app.yml:182]: use the `tests` module -- use sub commands to distinguish the task -- no raw python code in github actions

File: .github/workflows/python-app.yml (hunk 8 above)
Comment (JosuaKrause) [reply .github/workflows/python-app.yml:123]: are you overwriting results.xml?

File: tests/__main__.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
+import argparse
+
+from tests.utils import split_tests
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(
```
Comment (JosuaKrause) [inline tests/__main__.py:7]: use subcommands

File: tests/test_by_date.py [hunk 13]
```diff
@@ -73,7 +73,7 @@ def test_by_date_json(by_date: ByDate) -> None:
     start_date = "2022-01-03"
```
Comment (JosuaKrause) [inline tests/test_by_date.py:73]: if `tests/data_mini/data-2022.json` is mini then why is it too big for diff still?

File: tests/utils.py (hunk 4 above)
Comment (JosuaKrause) [reply tests/utils.py:96]: okay

File: .github/workflows/python-app.yml (hunk 8 above)
Comment (TheAnkurGoswami) [reply .github/workflows/python-app.yml:123]: I'm combining it in `coverage` workflow

Comment (TheAnkurGoswami) [reply .github/workflows/python-app.yml:123]: the result.xml will be overwritten only if we have any change in tests or the data

File: tests/test_by_date.py (hunk 13 above)
Comment (TheAnkurGoswami) [reply tests/test_by_date.py:73]: we have only 179 signals in it

File: .github/workflows/python-app.yml (hunk 9 above)
Comment (JosuaKrause) [reply .github/workflows/python-app.yml:96]: !

File: .github/workflows/python-app.yml [hunk 14]
```diff
@@ -105,15 +115,87 @@ jobs:
...
+    - name: Merge results
+      run: |
+        mv results-${{ matrix.python }}-*/results-${{ matrix.python }}-* test-results/parts/
+        python3 -m tests merge_results --dir test-results --out-fname results.xml
+
+    - name: Restore test-result Cache (Dummy step)  # This is a dummy step, i.e., we don't want to restore cache here. The post counterpart of this step would save our cache.
+      uses: actions/cache@v3
+      with:
+        path: test-results/*
+        key:  v1-test-results-${{ matrix.python }}-${{ hashFiles('tests/test*.py') }}-${{ hashFiles('tests/data*') }}
```
Comment (JosuaKrause) [inline .github/workflows/python-app.yml:188]: tests/data/**

File: tests/__main__.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+def parse_args_split_tests(parser: argparse.ArgumentParser) -> None:
+    parser.add_argument(
+        "--filepath",
+        default="test-results/results.xml",
+        type=str,
+        help=(
+            "The location for timings of all test cases. "
+            "Must be an xml file."))
+
```
Comment (JosuaKrause) [inline tests/__main__.py:14]: empty line?

Comment (JosuaKrause) [reply tests/__main__.py:14]:
```suggestion
```

**Review by JosuaKrause — APPROVED** [review]

**Pull request merged**